cpp/tests/sort/rank_test.cpp

```diff
@@ -0,0 +1,190 @@
+                    cudf::null_order null_precedence,
+                    bool print=false
+                    ) {
+    // Rank
+    auto got_rank_table = cudf::experimental::rank( input,
+                                                    method,
+                                                    column_order, 
+                                                    _include_nulls, 
+                                                    null_precedence,
+                                                    false);
+    if(print) {
+        cudf::test::print(got_rank_table->view().column(0)); std::cout<<"\n";
+    }
+    cudf::test::expect_tables_equal(expected, got_rank_table->view());
+}
+
+template <typename T>
+struct Rank : public cudf::test::BaseFixture {};
+
+TYPED_TEST_CASE(Rank, cudf::test::NumericTypes);
+
+TYPED_TEST(Rank, First)
+{
+    using T = TypeParam;
+
+    cudf::test::fixed_width_column_wrapper<T>   col1{{  5,   4,   3,   5,   8,   5}};
+    cudf::test::fixed_width_column_wrapper<T>   col2{{  5,   4,   3,   5,   8,   5}, {1, 1, 0, 1, 1, 1}};
+    cudf::test::strings_column_wrapper          col3({"d", "e", "a", "d", "k", "d"}, {1, 1, 1, 1, 1, 1});
+
+    //FIRST
+    //ASCENDING
+    cudf::test::fixed_width_column_wrapper<double> col1_asce_keep   {{3, 2, 1, 4, 6, 5}};
+    cudf::test::fixed_width_column_wrapper<double> col2_asce_keep   {{2, 1, 6, 3, 5, 4}, {1, 1, 0, 1, 1, 1}}; //KEEP
+    cudf::test::fixed_width_column_wrapper<double> col2_asce_top    {{3, 2, 1, 4, 6, 5}}; //BEFORE = TOP
+    cudf::test::fixed_width_column_wrapper<double> col2_asce_bottom {{2, 1, 6, 3, 5, 4}}; //AFTER  = BOTTOM
+    cudf::test::fixed_width_column_wrapper<double> col3_asce_keep   {{2, 5, 1, 3, 6, 4}, {1, 1, 1, 1, 1, 1}};
+    cudf::test::fixed_width_column_wrapper<double> col3_asce_keep2  {{2, 5, 1, 3, 6, 4}};
+    //DESCENDING
+    cudf::test::fixed_width_column_wrapper<double> col1_desc_keep   {{2, 5, 6, 3, 1, 4}};
+    cudf::test::fixed_width_column_wrapper<double> col2_desc_keep   {{3, 6, 1, 4, 2, 5}, {1, 1, 0, 1, 1, 1}}; //KEEP
+    cudf::test::fixed_width_column_wrapper<double> col2_desc_bottom {{2, 5, 6, 3, 1, 4}}; //BEFORE = BOTTOM
+    cudf::test::fixed_width_column_wrapper<double> col2_desc_top    {{3, 6, 1, 4, 2, 5}}; //AFTER  = TOP
+    cudf::test::fixed_width_column_wrapper<double> col3_desc_keep   {{3, 2, 6, 4, 1, 5}, {1, 1, 1, 1, 1, 1}};
+    cudf::test::fixed_width_column_wrapper<double> col3_desc_keep2  {{3, 2, 6, 4, 1, 5}};
+    auto first{cudf::rank_method::FIRST};
+
+    // Rank
+    if (std::is_same<T, cudf::experimental::bool8>::value) return;
+    // Single column, Ascending
+    //Non-null column
+    run_rank_test(  cudf::table_view{{col1}}, cudf::table_view{{col1_asce_keep}}, first,
+                    cudf::order::ASCENDING, 
+                    cudf::include_nulls::NO,  cudf::null_order::AFTER);
+    //Null column
+    run_rank_test(  cudf::table_view{{col2}}, cudf::table_view{{col2_asce_keep}}, first,
```

Comment:
This isn't how to write unit tests. Each of these needs to be it's own independent test.

Reply:
5 methods,  2 order, 3 na_options,null/non-null columns.

Reply:
CrossProduct?

Reply:
> too many tests to write:

You already wrote the tests, you just put them all in a single `TYPED_TEST`. They should be independent tests. That is the whole point of unit testing. 

None of our libcudf tests capture all of the possible combinations of the input space and I'm not asking you to do so. Just follow normal unit testing best practice and make each test independent. 

> CrossProduct

This only works for type lists.

Reply:
I added most of the input combinations. I separated tests with method.